File: swim-system-js/swim-core-js/@swim/streamlet/main/combinator/FilterFieldsOperator.ts
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Iterator, Cursor, Map} from "@swim/util";
import {BTree} from "@swim/collections";
import {KeyEffect} from "../KeyEffect";
import {AbstractMapInletMapOutlet} from "../AbstractMapInletMapOutlet";

export abstract class FilterFieldsOperator<K, V, I> extends AbstractMapInletMapOutlet<K, K, V, V, I, Map<K, V>> {
  has(key: K): boolean {
    if (this._input !== null) {
      const value = this._input.get(key);
      return value !== void 0 && this.evaluate(key, value);
    }
    return false;
  }

  get(): Map<K, V>;
  get(key: K): V | undefined;
  get(key?: K): Map<K, V> | V | undefined {
    if (key === void 0) {
      const output = new BTree<K, V>();
      const keys = this.keyIterator();
      do {
        const next = keys.next();
        if (!next.done) {
          const key = next.value!;
          const value = this._input!.get(key);
          if (value !== void 0 && this.evaluate(key, value)) {
            output.set(key, value);
          }
          continue;
        }
        break;
      } while (true);
      return output;
    } else {
      if (this._input !== null) {
        const value = this._input.get(key);
        if (value !== void 0 && this.evaluate(key, value)) {
          return value;
        }
      }
      return void 0;
    }
  }

  keyIterator(): Iterator<K> {
    if (this._input !== null) {
      return this._input.keyIterator(); // TODO: filter keys
    } else {
      return Cursor.empty();
    }
  }

  protected onInvalidateOutputKey(key: K, effect: KeyEffect): void {
    this.invalidateInputKey(key, effect);
  }

  protected onReconcileOutputKey(key: K, effect: KeyEffect, version: number): void {
    this.reconcileInputKey(key, version);
  }

  protected willReconcileInputKey(key: K, effect: KeyEffect, version: number): KeyEffect {
    if (effect === KeyEffect.Update) {
      if (this._input !== null) {
        const value = this._input.get(key);
        if (value === void 0 || !this.evaluate(key, value)) {
          return KeyEffect.Remove;
        }
      }
    }
    return effect;
  }

  abstract evaluate(key: K, value: V): boolean;
}

File: swim-system-java/swim-core-java/swim.streamlet/src/main/java/swim/streamlet/combinator/FilterFieldsOperator.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.streamlet.combinator;

import java.util.Iterator;
import java.util.Map;
import swim.collections.HashTrieMap;
import swim.streamlet.AbstractMapInletMapOutlet;
import swim.streamlet.KeyEffect;
import swim.util.Cursor;

public abstract class FilterFieldsOperator<K, V, I> extends AbstractMapInletMapOutlet<K, K, V, V, I, Map<K, V>> {
  @Override
  public boolean containsKey(K key) {
    if (this.input != null) {
      final V value = this.input.get(key);
      return value != null && evaluate(key, value);
    }
    return false;
  }

  @Override
  public V get(K key) {
    if (this.input != null) {
      final V value = this.input.get(key);
      if (value != null && evaluate(key, value)) {
        return value;
      }
    }
    return null;
  }

  @Override
  public Map<K, V> get() {
    HashTrieMap<K, V> output = HashTrieMap.empty();
    final Iterator<K> keys = keyIterator();
    while (keys.hasNext()) {
      final K key = keys.next();
      final V value = this.input.get(key);
      if (value != null && evaluate(key, value)) {
        output = output.updated(key, value);
      }
    }
    return output;
  }

  @Override
  public Iterator<K> keyIterator() {
    if (this.input != null) {
      return this.input.keyIterator();
    } else {
      return Cursor.empty();
    }
  }

  @Override
  protected void onInvalidateOutputKey(K key, KeyEffect effect) {
    invalidateInputKey(key, effect);
  }

  @Override
  protected void onReconcileOutputKey(K key, KeyEffect effect, int version) {
    reconcileInputKey(key, version);
  }

  @Override
  protected KeyEffect willReconcileInputKey(K key, KeyEffect effect, int version) {
    if (effect == KeyEffect.UPDATE) {
      if (this.input != null) {
        final V value = this.input.get(key);
        if (value == null || !evaluate(key, value)) {
          return KeyEffect.REMOVE;
        }
      }
    }
    return effect;
  }

  public abstract boolean evaluate(K key, V value);
}
